src/bit-systems/link-hover-menu.ts

```diff
@@ -1,47 +1,154 @@
+  const text = findChildWithComponent(world, TextTag, button)!;
+  const textObj = world.eid2obj.get(text)! as TroikaText;
+  const linkEid = LinkHoverMenu.targetObjectRef[menu];
+  const linkType = Link.type[linkEid];
+  let label = "";
+  switch (linkType) {
+    case LinkType.LINK:
+      label = "open link";
+      break;
+    case LinkType.AVATAR:
+      label = "use avatar";
+      break;
+    case LinkType.SCENE:
+      label = "use scene";
+      break;
+    case LinkType.WAYPOINT:
+      label = "go to";
+      break;
+    case LinkType.ROOM:
+    case LinkType.ROOM_URL:
+      label = "visit room";
+      break;
+  }
+  textObj.text = label;
+}
+
+function flushToObject3Ds(world: HubsWorld, menu: EntityID, frozen: boolean, force: boolean) {
+  const target = LinkHoverMenu.targetObjectRef[menu];
+  const visible = !!target && !frozen;
 
-  // Check if the cursor it hovered on Link object or Link menu button object.
-  const hovered = hoveredQuery(world).length > 0 || hoveredMenuItemQuery(world).length > 0;
-
-  const linkEid = LinkHoverMenu.targetObjectRef[menuEid];
-  if (hovered && entityExists(world, linkEid)) {
-    // Hovered then make the menu visible and handle clicks if needed.
-    menuObject.visible = true;
-    if (clickedMenuItemQuery(world).length > 0) {
-      // TODO: Change the action and label depending on 
-      //       target (eg: switching avatars for avatar url)
-      window.open(APP.getString(Link.url[linkEid])!);
+  const obj = world.eid2obj.get(menu)!;
+  obj.visible = visible;
+
+  const linkButtonRef = LinkHoverMenu.linkButtonRef[menu];
+  const buttonObj = world.eid2obj.get(linkButtonRef)!;
+  // Parent visibility doesn't block raycasting, so we must set each button to be invisible
+  // TODO: Ensure that children of invisible entities aren't raycastable
+  if (visible) {
+    if (visible !== buttonObj.visible || force) {
+      updateButtonText(world, menu, linkButtonRef);
+      buttonObj.visible = true;
     }
   } else {
-    // Not hovered or target object has already been deleted
-    // then make the menu invisible and forget the Link object.
-    menuObject.visible = false;
-    if (menuObject.parent !== null) {
-      menuObject.parent.remove(menuObject);
-    }
-    LinkHoverMenu.targetObjectRef[menuEid] = NULL_EID;
+    buttonObj.visible = false;
   }
 }
+
+export function linkHoverMenuSystem(world: HubsWorld, sceneIsFrozen: boolean) {
+  // Assumes always only single LinkHoverMenu entity exists for now.
+  // TODO: Take into account for more than one for VR
+  menuQuery(world).forEach(menu => {
+    const prevTarget = LinkHoverMenu.targetObjectRef[menu];
+    updateLinkMenuTarget(world, menu, sceneIsFrozen);
+    const currTarget = LinkHoverMenu.targetObjectRef[menu];
+    if (currTarget) {
+      moveToTarget(world, menu);
+      clickedMenuItemQuery(world).forEach(eid => handleLinkClick(world, eid));
+    }
+    flushToObject3Ds(world, menu, sceneIsFrozen, prevTarget != currTarget);
```

Comment:
Let's use `!==` instead of `!=` for the consistency.

```suggestion
    flushToObject3Ds(world, menu, sceneIsFrozen, prevTarget !== currTarget);
```